package/src/error.rs

```diff
@@ -0,0 +1,188 @@
+    TempFilePersist {
+        error: tempfile::PersistError,
+    },
+    NoManifestParent,
+    /// Index dependencies aren't implemented yet, so we emit
+    /// this if we encounter one.
+    IndexDep,
+}
+
+impl std::error::Error for Error {}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::Io { error, path } => {
+                if let Some(path) = path {
+                    write!(f, "{}: {error}", path.display())
+                } else {
+                    error.fmt(f)
+                }
+            }
+            // Just a short and not-very-informative error. To write a better error message to
+            // the terminal, use `program.report` like the cli does.
+            Error::ManifestEval { package, .. } => {
+                if let Some(package) = package {
+                    write!(f, "error evaluating manifest for package {package}")
+                } else {
+                    write!(f, "error evaluating package manifest")
+                }
+            }
+            Error::RestrictedPath {
+                attempted,
+                restriction,
+                package_url,
+                package_commit,
+                package_path,
+            } => {
+                write!(
+                    f,
+                    "git package {package_url}@{package_commit}/{} tried to import path {}, but can only import from {}",
+                    package_path.display(),
+                    attempted.display(),
+                    restriction.display()
+                )
+            }
+            Error::Git(e) => e.fmt(f),
+            Error::InvalidUrl { msg, .. } => {
+                write!(f, "{}", msg)
+            }
+            Error::InternalManifestError { path, msg } => {
+                write!(
+                    f,
+                    "internal error reading the manifest at {}; this is a bug in nickel: {msg}",
+                    path.display()
+                )
+            }
+            Error::NoPackageRoot { path } => write!(
+                f,
+                "tried to import a relative path ({}), but we have no reference",
+                path.display()
+            ),
+            Error::Resolution { msg } => write!(f, "version resolution failed: {msg}"),
+            Error::TempFilePersist { error } => error.fmt(f),
+            Error::NoProjectDir => {
+                write!(
+                    f,
+                    "failed to find a location for the nickel cache directory"
+                )
+            }
+            Error::LockFileDeserialization { path, error } => {
+                write!(f, "lock file {} is invalid: {error}", path.display())
+            }
+            Error::NoManifestParent => {
```

Comment:
What does this  error correspond to? Is that possible for a regular file to have no parent?

Reply:
I think only empty paths have no parent, so I've dropped this error and just added a panic in the manifest constructor.